libs/contrib/Data/Nat.idr

```diff
@@ -16,10 +31,83 @@ ltePlus (LTESucc lte1) lte2 = LTESucc $ ltePlus lte1 lte2
 lteCongPlus : (k : Nat) -> LTE m n -> LTE (m + k) (n + k)
 lteCongPlus k lte = ltePlus lte lteRefl
 
+lteMultRight : LTE a b -> (c : Nat) -> LTE a (b * S c)
+lteMultRight {b} prf c =
+        rewrite multRightSuccPlus b c in
+        lteTransitive prf $ lteAddRight b
+
 lteMult : LTE m1 n1 -> LTE m2 n2 -> LTE (m1 * m2) (n1 * n2)
 lteMult LTEZero _ = LTEZero
 lteMult {m1=S k} (LTESucc _) LTEZero = rewrite multZeroRightZero k in LTEZero
 lteMult (LTESucc lte1) (LTESucc lte2) = LTESucc $ ltePlus lte2 $ lteMult lte1 $ LTESucc lte2
 
 lteCongMult : (k : Nat) -> LTE m n -> LTE (m * k) (n * k)
 lteCongMult k lte = lteMult lte lteRefl
+
+addLteRight : LTE a b -> LTE (a + c) (b + c)
+addLteRight {a} {b} {c = Z} prf =
+        rewrite plusZeroRightNeutral a in
+        rewrite plusZeroRightNeutral b in
+        prf
+addLteRight {a} {b} {c = S k} prf =
+        rewrite sym $ plusSuccRightSucc a k in
+        rewrite sym $ plusSuccRightSucc b k in
+        LTESucc $ addLteRight prf
+
+addLteLeft : LTE b c -> LTE (a + b) (a + c)
+addLteLeft {a = Z} {b} {c} prf = prf
+addLteLeft {a = S k} {b} {c} prf =
+        LTESucc $ addLteLeft prf
+
+subtractLteLeft : LTE (a + b) (a + c) -> LTE b c
+subtractLteLeft {a = Z} {b} {c} prf = prf
+subtractLteLeft {a = S j} {b} {c} prf = subtractLteLeft $ fromLteSucc prf
+
+subtractEqLeft : {a, b, c : Nat} -> a + b = a + c -> b = c
+subtractEqLeft {a = Z} prf = prf
+subtractEqLeft {a = S k} {b} {c} prf = subtractEqLeft $ succInjective (k + b) (k + c) prf
+
+subtractEqRight : {a, b, c : Nat} -> a + c = b + c -> a = b
```

Comment:
`subtractEqRight` is can be proved as an instance of `subtractEqLeft`:

```idris
subtractEqRight : {a, b, c : Nat} -> a + c = b + c -> a = b
subtractEqRight {a} {b} {c} prf =
  subtractEqLeft {a = c} {b = a} {c = b} $
    trans (plusCommutative c a) $
      trans prf $
        plusCommutative b c
```

Reply:
Good point!